### PR TITLE
Add total_bounds of layer to `read_info`

### LINKED ISSUE #274: Get dataset bounds via `pyogrio.read_info`
Is it possible to access a dataset's spatial extent for supported drivers without fetch all the features with `read_bounds`? Some drivers like FlatGeobuf know this in advance. E.g. if I run

```
ogrinfo /vsicurl/https://data.source.coop/cholmes/eurocrops/unprojected/flatgeobuf/FR_2018_EC21.fgb FR_2018_EC21 -so
```
then I see 
```
...
Extent: (-6.047022, -3.916365) - (68.890504, 51.075101)
...
```
without waiting for the 6.5GB file to download. Can we expose that information through `pyogrio.read_info`?

### PR DESCRIPTION
**Pull request by theroggy**
I made `force_feature_count` default to `False`, which is a breaking change for eg. geojson: will return -1 for `read_info()["features"]` now by default because determining number features is not fast for this file type. 
Not sure if this is OK? If OK, I'll add it as a breaking change in the changelog.

closes #274

### CODE REVIEW

**Review by brendan-ward — COMMENTED** [review]
Thanks @theroggy !

What do you think about exposing a couple items to the capabilities list as well as [suggested in 274](https://github.com/geopandas/pyogrio/issues/274#issuecomment-1687281701):
* `fast_total_bounds`: `OGR_L_TestCapability(ogr_layer, OLCFastGetExtent)`
* `fast_feature_count`: `OGR_L_TestCapability(ogr_layer, OLCFastFeatureCount)`


I'm OK with adding a breaking change for counts for GeoJSON data sources for `read_info`; this keeps us closer in parity to the underlying driver's capabilities.

File: pyogrio/_io.pyx [hunk 1]
```diff
@@ -376,6 +378,33 @@ cdef get_feature_count(OGRLayerH ogr_layer):
...
+
+    Parameters
+    ----------
+    ogr_layer : pointer to open OGR layer
+    force : bool
+        True if the total bounds should be computed even if it is expensive
+
+    Returns
+    -------
+    Optional[Tuple[float, float, float, float]]
```
[the remark this answers is not in the document]
Comment (theroggy) [reply pyogrio/_io.pyx:392]: ok

File: pyogrio/core.py [hunk 2]
```diff
@@ -148,11 +148,25 @@ def read_bounds(
...
 
-    ``crs`` and ``geometry`` will be ``None`` and ``features`` will be 0 for a
-    nonspatial layer.
+    ``crs``, ``geometry`` and ``total_bounds`` will be ``None`` and ``features`` will be
+    0 for a nonspatial layer.
+
+    ``features`` will be -1 if determining this is expensive. You can force to calculate
+    it anyway using the ``force_featurecount`` parameter.
+
+    ``total_bounds`` will be None if determining this is expensive. You can force to
```
[the remark this answers is not in the document]
Comment (kylebarron) [reply pyogrio/core.py:168]: You probably want `if this is **an** expensive`

File: pyogrio/tests/test_core.py [hunk 3]
```diff
@@ -268,19 +270,51 @@ def test_read_info_invalid_dataset_kwargs(naturalearth_lowres):
 
 def test_read_info_force_feature_count_exception(data_dir):
     with pytest.raises(DataLayerError, match="Could not iterate over features"):
-        read_info(data_dir / "sample.osm.pbf", layer="lines")
+        read_info(data_dir / "sample.osm.pbf", layer="lines", force_featurecount=True)
 
 
-def test_read_info_force_feature_count(data_dir):
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (theroggy) [reply pyogrio/tests/test_core.py:276]: Yes, this wans't a very pretty test...

**Comment by theroggy** [comment]
> What do you think about exposing a couple items to the capabilities list as well as [suggested in 274](https://github.com/geopandas/pyogrio/issues/274#issuecomment-1687281701):
> 
> * `fast_total_bounds`: `OGR_L_TestCapability(ogr_layer, OLCFastGetExtent)`
> * `fast_feature_count`: `OGR_L_TestCapability(ogr_layer, OLCFastFeatureCount)`

Oops, forgot that one... Added them. Apparently the capabilities there had value 0 or 1, which isn't ideal I think? So I changed them to boolean.

> I'm OK with adding a breaking change for counts for GeoJSON data sources for `read_info`; this keeps us closer in parity to the underlying driver's capabilities.

Done. Also added that the capabilities are now booleans.

File: pyogrio/_io.pyx [hunk 4]
```diff
@@ -376,6 +378,33 @@ cdef get_feature_count(OGRLayerH ogr_layer):
...
+    force : bool
+        True if the total bounds should be computed even if it is expensive
+
+    Returns
+    -------
+    tuple of (xmin, ymin, xmax, ymax) or None
+        The total bounds of the layer, or None if they could not be determined.
+    """
+
+    cdef OGREnvelope ogr_envelope # = NULL
```
Comment (brendan-ward) [inline pyogrio/_io.pyx:396]:
```suggestion
    cdef OGREnvelope ogr_envelope = NULL
```

File: pyogrio/_io.pyx [hunk 5]
```diff
@@ -376,6 +378,33 @@ cdef get_feature_count(OGRLayerH ogr_layer):
...
+    try:
+        exc_wrap_ogrerr(OGR_L_GetExtent(ogr_layer, &ogr_envelope, force))
+        bounds = (
```
Comment (brendan-ward) [inline pyogrio/_io.pyx:399]: I imagine there could be a slight though unlikely chance that `ogr_envelope` remains `NULL` without an error being raised, so I'd suggest a check for that before extracting bounds:

```
if ogr_envelope == NULL:
    bounds = None
else:
    bounds = ...

```

Comment (theroggy) [reply pyogrio/_io.pyx:399]: ok

**Review by jorisvandenbossche — APPROVED** [review]
Looks good to me!

**Review by brendan-ward — APPROVED** [review]

**Pull request merged**